source/extensions/common/wasm/wasm.cc

```diff
@@ -263,16 +276,24 @@ static void createWasmInternal(const VmConfig& vm_config, const PluginSharedPtr&
         complete_cb(code);
       }
       // NB: must be deleted explicitly.
-      dispatcher.deferredDelete(
-          Envoy::Event::DeferredDeletablePtr{remote_data_provider_holder->release()});
-      remote_data_provider_holder->reset();
+      if (*holder) {
+        dispatcher.deferredDelete(Envoy::Event::DeferredDeletablePtr{holder->release()});
+      }
     };
-    remote_data_provider = std::make_unique<Config::DataSource::RemoteAsyncDataProvider>(
-        cluster_manager, init_manager, vm_config.code().remote(), dispatcher, random, true,
-        fetch_callback);
     if (fail_if_code_not_cached) {
-      *remote_data_provider_holder = std::move(remote_data_provider);
+      auto adapter = std::make_unique<RemoteDataFetcherAdapter>(fetch_callback);
+      auto fetcher = std::make_unique<Config::DataFetcher::RemoteDataFetcher>(
+          cluster_manager, vm_config.code().remote().http_uri(), vm_config.code().remote().sha256(),
+          *adapter);
+      auto fetcher_ptr = fetcher.get();
+      adapter->setFetcher(std::move(fetcher));
+      *holder = std::move(adapter);
+      fetcher_ptr->fetch();
```

Comment:
Using pointer that's moved (twice!) seems like an anti-pattern. Is there any way to make this a bit cleaner?

Reply:
The problem here is that we need to make the  fetcher self-deleting.  We also need to use an object for the callback because that is the API for the fetcher.   Sooooo, that means creating a callback object creating the fetcher.  Storing the fetcher into the callback object, then passing the callback object into the callback function as a shared pointer.   I could switch to bare pointers which would simplify things a bit.  WDYT?  The core problem is that the API for this fetcher is not very flexible.  It is really meant to be used in the single existing use-case.

Reply:
I could also make the function into a member function.  That would be less clear since the callback code is logically part of the createWasmInternal function, but it would eliminate some unique_ptr movement.

Reply:
Aren't we the only existing use-case (i.e. couldn't we change the API for the fetcher)?

Reply:
We could.  That would require upstreaming to envoy.  I am absolutely willing to do that, but this is being backported to 1.6 so I didn't want to make changes into core envoy as part of the backport.  We can discuss this offline.  I agree that in the long run we should change the fetcher API.  The question is what to do in the short term.

Reply:
Fair enough, let's merge this as-is. Could you open tracking issue for that?

Reply:
https://github.com/envoyproxy/envoy-wasm/issues/506